Source/player.cpp

```diff
@@ -28,6 +28,189 @@
+	y = (y / 8) * (y < 0 ? 1 : -1);
+	int lx = x + (l->position.tile.x * 8);
+	int ly = y + (l->position.tile.y * 8);
+	int offx = l->position.offset.x + (l->position.tile.x * 8);
+	int offy = l->position.offset.y + (l->position.tile.y * 8);
+
+	if (abs(lx - offx) < 3 && abs(ly - offy) < 3)
+		return;
+
+	ChangeLightOff(player._plid, { x, y });
+}
+
+void WalkUpwards(int pnum, Point vel, DirectionSettings const &walkParams)
+{
+	auto &player = plr[pnum];
+	dPlayer[player.position.future.x][player.position.future.y] = -(pnum + 1);
+	player.position.temp = walkParams.tileAdd;
+}
+
+void WalkDownwards(int pnum, Point vel, DirectionSettings const &walkParams)
+{
+	auto &player = plr[pnum];
+	dPlayer[player.position.tile.x][player.position.tile.y] = -(pnum + 1);
+	player.position.temp = player.position.tile;
+	player.position.tile = player.position.future; // Move player to the next tile to maintain correct render order
+	dPlayer[player.position.tile.x][player.position.tile.y] = pnum + 1;
+	ChangeLightXY(player._plid, player.position.tile);
+	PM_ChangeLightOff(player);
+}
+
+void WalkSides(int pnum, Point vel, DirectionSettings const &walkParams)
+{
+	auto &player = plr[pnum];
+
+	Point const nextPosition = walkParams.map + player.position.tile;
+
+	dPlayer[player.position.tile.x][player.position.tile.y] = -(pnum + 1);
+	dPlayer[player.position.future.x][player.position.future.y] = -(pnum + 1);
+	player._pVar4 = nextPosition.x;
+	player._pVar5 = nextPosition.y;
+	dFlags[nextPosition.x][nextPosition.y] |= BFLAG_PLAYERLR;
+
+	if (leveltype != DTYPE_TOWN) {
+		ChangeLightXY(player._plid, nextPosition);
+		PM_ChangeLightOff(player);
+	}
+
+	player.position.temp = player.position.future;
+}
+
+static constexpr std::array<DirectionSettings const, 8> directionSettings { { { DIR_S, { 1, 1 }, { 0, -32 }, { 0, 0 }, SDIR_S, PM_WALK2, WalkDownwards },
+	{ DIR_SW, { 0, 1 }, { 32, -16 }, { 0, 0 }, SDIR_SW, PM_WALK2, WalkDownwards },
+	{ DIR_W, { -1, 1 }, { 32, -16 }, { 0, 1 }, SDIR_W, PM_WALK3, WalkSides },
+	{ DIR_NW, { -1, 0 }, { 0, 0 }, { 0, 0 }, SDIR_NW, PM_WALK, WalkUpwards },
+	{ DIR_N, { -1, -1 }, { 0, 0 }, { 0, 0 }, SDIR_N, PM_WALK, WalkUpwards },
+	{ DIR_NE, { 0, -1 }, { 0, 0 }, { 0, 0 }, SDIR_NE, PM_WALK, WalkUpwards },
+	{ DIR_E, { 1, -1 }, { -32, -16 }, { 1, 0 }, SDIR_E, PM_WALK3, WalkSides },
+	{ DIR_SE, { 1, 0 }, { -32, -16 }, { 0, 0 }, SDIR_SE, PM_WALK2, WalkDownwards } } };
+
+void ScrollViewPort(int pnum, _scroll_direction dir)
+{
+	auto const &player = plr[pnum];
+
+	if (pnum != myplr) {
+		return;
+	}
+
+	ScrollInfo.tile.x = player.position.tile.x - ViewX;
+	ScrollInfo.tile.y = player.position.tile.y - ViewY;
+
+	if (zoomflag) {
+		if (abs(ScrollInfo.tile.x) >= 3 || abs(ScrollInfo.tile.y) >= 3) {
+			ScrollInfo._sdir = SDIR_NONE;
+		} else {
+			ScrollInfo._sdir = dir;
+		}
+	} else if (abs(ScrollInfo.tile.x) >= 2 || abs(ScrollInfo.tile.y) >= 2) {
+		ScrollInfo._sdir = SDIR_NONE;
+	} else {
+		ScrollInfo._sdir = dir;
+	}
+}

```

Comment:
Looking at this again I think it would be better to rewrite it as:
```suggestion
void ScrollViewPort(_scroll_direction dir)
{
	auto const &myPlayer = plr[myplr];

	ScrollInfo.tile.x = myPlayer.position.tile.x - ViewX;
	ScrollInfo.tile.y = myPlayer.position.tile.y - ViewY;

	int distance = 2;
	if (zoomflag) {
        distance++;
	}
	ScrollInfo._sdir = SDIR_NONE;
	if (abs(ScrollInfo.tile.x) < distance && abs(ScrollInfo.tile.y) < distance) {
		ScrollInfo._sdir = dir;
	}
}
```

And then change the call site to:
```cpp
if (pnum == myplr) {
    ScrollViewPort(directionSettings[Dir].scrollDir);
}
```

Reply:
Sorted